hazelcast/src/main/java/com/hazelcast/map/impl/querycache/subscriber/SubscriberAccumulatorHandler.java

```diff
@@ -91,61 +92,73 @@ public void handle(QueryCacheEventData eventData, boolean ignored) {
-
-        if (!hasMissingCount(removedEntryCounts)) {
-            if (removedEntryCounts.compareAndSet(partitionCount, POLL_PERMIT, null)) {
-                try {
-                    int totalRemovedEntryCount = pollRemovedEntryCounts(removedEntryCounts);
-                    publishCacheWideEvent(((DefaultQueryCache) queryCache).context,
-                            ((DefaultQueryCache) queryCache).mapName,
-                            ((DefaultQueryCache) queryCache).cacheId,
-                            totalRemovedEntryCount,
-                            eventType);
-                } finally {
-                    removedEntryCounts.set(partitionCount, POLL_PERMIT);
+        removedCountHolders.get(partitionId).offer(removedEntryCount);
+
+        if (noMissingMapWideEvent(removedCountHolders)
+                && removedCountHolders.compareAndSet(partitionCount, POLL_PERMIT, null)) {
+            try {
+                if (noMissingMapWideEvent(removedCountHolders)) {
+                    int totalRemovedCount = pollRemovedCountHolders(removedCountHolders);
+                    publishCacheWideEvent(queryCache, totalRemovedCount, eventType);
                 }
+            } finally {
+                removedCountHolders.set(partitionCount, POLL_PERMIT);
             }
         }
     }
 
     /**
-     * @return {@code true} if we have removed entries from all partitions,
-     * otherwise return {@code false} to indicate there is still
-     * not-received events for some partitions.
+     * Check if all map-wide events like {@link
+     * EntryEventType#CLEAR_ALL} or {@link EntryEventType#EVICT_ALL}
+     * were received. If an event is received, we populate its
+     * partitions' removed-entry-count-holder queue.
+     *
+     * @return {@code true} if we have received map-wide events from all
+     * partitions, otherwise return {@code false} to indicate there is
+     * still not-received map-wide events for some partitions.
      */
-    private boolean hasMissingCount(AtomicReferenceArray<Queue<Integer>> removedEntryCounts) {
+    private boolean noMissingMapWideEvent(AtomicReferenceArray<Queue<Integer>> removedCountHolders) {
         for (int i = 0; i < partitionCount; i++) {
-            if (removedEntryCounts.get(i).size() < 1) {
-                // we don't receive any map-wide event for this partition
-                return true;
+            if (removedCountHolders.get(i).size() < 1) {
+                // means we still have not-received map-wide event for this partition
+                return false;
             }
         }
-        return false;
+        return true;
     }
 
-    // should be called when `hasMissingCount` `false`, otherwise polling can cause NPE
-    private int pollRemovedEntryCounts(AtomicReferenceArray<Queue<Integer>> removedEntryCounts) {
+    // should be called when `hasMissingMapWideEvent` `false`, otherwise polling can cause NPE
```

Comment:
Leftover reference to `hasMissingMapWideEvent` before its logic was reversed as `noMissingMapWideEvent`?